src/Functions/IFunction.cpp

```diff
@@ -670,6 +671,37 @@ FunctionBasePtr IFunctionOverloadResolver::build(const ColumnsWithTypeAndName &
         }
     }
 
+    /// Use FunctionBaseVariantAdaptor if default implementation for Variant is enabled and we have Variant type in arguments.
+    if (useDefaultImplementationForVariant())
+    {
+        checkNumberOfArguments(arguments.size());
+
+        /// Special case: if there are MULTIPLE arguments and ALL of them are Variants,
+        /// don't use VariantAdaptor. This allows functions (especially comparisons) to work
+        /// directly on ColumnVariant objects, which have built-in logic for cross-variant
+        /// comparisons using discriminator ordering.
+        bool has_variant = false;
+        bool all_variants = true;
+        for (const auto & arg : arguments)
+        {
+            if (isVariant(arg.type))
+                has_variant = true;
+            else
+                all_variants = false;
+        }
+
+        /// Only skip VariantAdaptor if:
+        /// 1. We have Variant arguments
+        /// 2. Either not all are Variants, OR there's only one argument (single Variant needs unwrapping)
+        if (has_variant && (!all_variants || arguments.size() == 1))
+        {
+            DataTypes data_types(arguments.size());
+            for (size_t i = 0; i < arguments.size(); ++i)
+                data_types[i] = arguments[i].type;
+            return std::make_shared<FunctionBaseVariantAdaptor>(shared_from_this(), std::move(data_types));
+        }
```

Comment:
Let's use default implementation for Variant in all cases when we have Variant argument. For functions that want to work with Variant directly we can set `useDefaultImplementationForVariant` to false.

For comparison functions, I don't think someone really relies on the built-in comparison rule, it was added just to make this type sortable. But just in case we can add a setting that enables/disables default implementation for Variant for compatibility. We can check it here using `CurrentThread::get().getQueryContext()` (not the best approach, but better than nothing).